cmd/crank/beta/validate/image.go

```diff
@@ -37,8 +39,56 @@ type Fetcher struct{}
 
 // FetchBaseLayer fetches the base layer of the image which contains the 'package.yaml' file.
 func (f *Fetcher) FetchBaseLayer(image string) (*conregv1.Layer, error) {
-	if strings.Contains(image, "sha") { // Strip the digest before fetching the image
-		image = strings.Split(image, "@")[0]
+	if strings.Contains(image, "@") {
+		// Strip the digest before fetching the image
+		image = strings.SplitN(image, "@", 2)[0]
+	} else if strings.Contains(image, ":") {
+		// Separate the image base and the image tag
+		parts := strings.SplitN(image, ":", 2)
+		imageBase := parts[0]
+		imageTag := parts[1]
+
+		// Check if the tag is a constraint
+		isConstraint := true
+		c, err := semver.NewConstraint(imageTag)
+		if err != nil {
+			isConstraint = false
+		}
+
+		if isConstraint {
+			// Fetch all image tags
+			tags, err := crane.ListTags(imageBase)
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot fetch tags for the image %s", imageBase)
+			}
+
+			// Convert tags to semver versions
+			vs := []*semver.Version{}
+			for _, r := range tags {
+				v, err := semver.NewVersion(r)
+				if err != nil {
+					// We skip any tags that are not valid semantic versions
+					continue
+				}
+				vs = append(vs, v)
+			}
+
+			// Sort all versions and find the last version complient with the constraint
+			sort.Sort(semver.Collection(vs))
+			var addVer string
+			for _, v := range vs {
+				if c.Check(v) {
+					addVer = v.Original()
+				}
+			}
+
+			if addVer == "" {
+				return nil, errors.Wrapf(err, "cannot find any tag complient with the constraint %s", imageTag)
+			}
+
+			// Compose new complete image string if any complient version was found
+			image = fmt.Sprintf("%s:%s", imageBase, addVer)
+		}
```

Comment:
Would it make sense to split this logic out into a separate function?

Reply:
Done in the last commit. Please feel free to review it.